packages/omgidl-serialization/src/MessageReader.ts

```diff
@@ -166,86 +166,111 @@ export class MessageReader<T = unknown> {
-            return deser(reader, arrayLengths[arrayLengths.length - 1]!);
-          };
-
-          // last arrayLengths length is handled in deserializer. It returns an array
-          return makeNestedArray(typedArrayDeserializer, arrayLengths.slice(0, -1), 0);
+
+      if (field.typeDeserInfo.type === "struct" || field.typeDeserInfo.type === "union") {
+        if (field.isArray === true) {
+          // For dynamic length arrays we need to read a uint32 prefix
+          const arrayLengths = field.arrayLengths ?? [reader.sequenceLength()];
+          return this.readComplexNestedArray(
+            reader,
+            childOptions,
+            field.typeDeserInfo,
+            arrayLengths,
+            0,
+          );
         } else {
-          return deser(reader, arrayLengths[0]!);
+          return this.readAggregatedType(
+            field.typeDeserInfo,
+            reader,
+            childOptions,
+            emHeaderSizeBytes,
+          );
         }
-      } else if (field.typeDeserInfo.type === "primitive") {
-        return field.typeDeserInfo.deserialize(
-          reader,
-          headerSpecifiedLength, // fieldLength only used for `string` type primitives
-        );
       } else {
-        throw new Error(`Unhandled deserialization info type`);
+        const headerSpecifiedLength =
+          emHeaderSizeBytes != undefined
+            ? Math.floor(emHeaderSizeBytes / field.typeDeserInfo.typeLength)
+            : undefined;
+
+        if (field.typeDeserInfo.type === "array-primitive") {
+          const deser = field.typeDeserInfo.deserialize;
+          const arrayLengths =
+            field.arrayLengths ??
+            // the byteLength written in the header doesn't help us determine count of strings in the array
+            // This will be the next field in the message
+            (field.type === "string"
+              ? [reader.sequenceLength()]
+              : [headerSpecifiedLength ?? reader.sequenceLength()]);
+
+          if (arrayLengths.length > 1) {
+            const typedArrayDeserializer = () => {
+              return deser(reader, arrayLengths[arrayLengths.length - 1]!);
+            };
+
+            // last arrayLengths length is handled in deserializer. It returns an array
+            return makeNestedArray(typedArrayDeserializer, arrayLengths.slice(0, -1), 0);
+          } else {
+            return deser(reader, arrayLengths[0]!);
+          }
+        } else if (field.typeDeserInfo.type === "primitive") {
+          return field.typeDeserInfo.deserialize(
+            reader,
+            headerSpecifiedLength, // fieldLength only used for `string` type primitives
+          );
+        } else {
+          throw new Error(`Unhandled deserialization info type`);
+        }
+      }
+    } catch (err: unknown) {
+      if (err instanceof Error) {
+        err.message = `${err.message} in field ${field.name} of ${
+          emHeaderOptions.parentName ?? ""
```

Comment:
nit: the error message might look weird if this parentName is missing - when would this happen & should we say something more specific like "unknown parent" (but that seems like it should maybe never happen?)

Reply:
good callout, updated the parameter to be required and udpated the callers to pass standins for the parentName allowing for `<unnamed-struct>` and `<unnamed-union>` parent names